data/market.py

```diff
@@ -0,0 +1,176 @@
+        """
+        self._base_token_amount -= amount / 2 / self.exchange_rate()
+        self._quote_token_amount -= amount / 2
+        self.set_invariant()
+
+    # TODO: refacture swap and calculate_slippage!
+
+    def exact_output_swap(self, output_token: Token, amount: int) -> float:
+        """Executes an exact output swap.
+
+        Args:
+            amount (int): Amount of 'output_token' to swap (buy from the pool)
+            output_token (Token): The token to receive as the output of the swap.
+
+        Raises:
+            Exception: Raises and error if swap amount is too large.
+
+        Returns:
+            float: _description_
+        """
+        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
+            token=output_token, amount= amount, type="exact_output"
+        )
+
+    def exact_input_swap(self, input_token: Token, amount: int) -> float:
+        """Executes and exacpt input swap.
+
+        Args:
+            amount (int): Amount of 'input_token' to swap (sell to the pool)
+            input_token (Token): The token to send to the pool as the input of the swap.
+
+        Returns:
+            float: _description_
+        """
+        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
+            token=input_token, amount=amount, type="exact_input"
+        )
+
+    def exact_input_swap_slippage(self, input_token: Token, amount: int) -> float:
+        """Executes and exacpt input swap.
+
+        Args:
+            amount (int): Amount of 'input_token' to swap (sell to the pool)
+            input_token (Token): The token to send to the pool as the input of the swap.
+
+        Returns:
+            float: _description_
+        """
+        _, _, slippage = self.calculate_params(
+            token=input_token, amount=amount, type="exact_input"
+        )
+        return slippage
+
+    def exact_output_swap_slippage(self, output_token: Token, amount: int) -> float:
+        """Executes an exact output swap.
+
+        Args:
+            amount (int): Amount of 'output_token' to swap (buy from the pool)
+            output_token (Token): The token to receive as the output of the swap.
+
+        Raises:
+            Exception: Raises and error if swap amount is too large.
+
+        Returns:
+            float: _description_
+        """
+        _, _, slippage = self.calculate_params(
+            token=output_token, amount=amount, type="exact_output"
+        )
+        return slippage
+
+    def calculate_params(self, token: Token, amount: int, type: str = "exact_output") -> tuple[float, float, float]:
+        if (token.name == self.base_token.name) & (type == "exact_output"):
+            # does the swap with the base token as output
+            if amount < self._base_token_amount:
+                _current_exchange_rate = self.exchange_rate()
```

Comment:
this line can be moved out of all the if-else branches to the end of the method, to be used when computing `_slippage`, right?

Reply:
Yep,done.